pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java

```diff
@@ -0,0 +1,256 @@
+      return _jsonMatchOperator.getTrues();
+    } else {
+      return _expressionFilterOperator.getTrues();
+    }
+  }
+
+  @Override
+  public boolean canOptimizeCount() {
+    if (_jsonMatchOperator != null) {
+      return _jsonMatchOperator.canOptimizeCount();
+    } else {
+      return _expressionFilterOperator.canOptimizeCount();
+    }
+  }
+
+  @Override
+  public int getNumMatchingDocs() {
+    if (_jsonMatchOperator != null) {
+      return _jsonMatchOperator.getNumMatchingDocs();
+    } else {
+      return _expressionFilterOperator.getNumMatchingDocs();
+    }
+  }
+
+  @Override
+  public boolean canProduceBitmaps() {
+    if (_jsonMatchOperator != null) {
+      return _jsonMatchOperator.canProduceBitmaps();
+    } else {
+      return _expressionFilterOperator.canProduceBitmaps();
+    }
+  }
+
+  @Override
+  public BitmapCollection getBitmaps() {
+    if (_jsonMatchOperator != null) {
+      return _jsonMatchOperator.getBitmaps();
+    } else {
+      return _expressionFilterOperator.getBitmaps();
+    }
+  }
+
+  @Override
+  public List<Operator> getChildOperators() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public String toExplainString() {
+    StringBuilder stringBuilder =
+        new StringBuilder(EXPLAIN_NAME).append("(column:").append(_columnName).append(",key:").append(_keyName)
+            .append(",indexLookUp:map_index").append(",operator:").append(_predicate.getType()).append(",predicate:")
+            .append(_predicate);
+
+    if (_jsonMatchOperator != null) {
+      stringBuilder.append(",delegateTo:json_match");
+    } else {
+      stringBuilder.append(",delegateTo:expression_filter");
+    }
+
+    return stringBuilder.append(')').toString();
+  }
+
+  @Override
+  protected String getExplainName() {
+    return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, EXPLAIN_NAME);
+  }
+
+  @Override
+  protected void explainAttributes(ExplainAttributeBuilder attributeBuilder) {
+    super.explainAttributes(attributeBuilder);
+    attributeBuilder.putString("column", _columnName);
+    attributeBuilder.putString("key", _keyName);
+    attributeBuilder.putString("indexLookUp", "map_index");
+    attributeBuilder.putString("operator", _predicate.getType().name());
+    attributeBuilder.putString("predicate", _predicate.toString());
+
+    if (_jsonMatchOperator != null) {
+      attributeBuilder.putString("delegateTo", "json_match");
+    } else {
+      attributeBuilder.putString("delegateTo", "expression_filter");
+    }
+  }
+
+  /**
+   * Initializes the JsonMatchFilterOperator with the given JsonIndexReader
+   */
+  private JsonMatchFilterOperator initializeJsonMatchFilterOperator(JsonIndexReader jsonIndex,
+      JsonMatchPredicate jsonMatchPredicate, int numDocs) {
+    return new JsonMatchFilterOperator(jsonIndex, jsonMatchPredicate, numDocs);
+  }
+
+  private String createJsonEqPredicateValue(String key, String value) {
+    return String.format("%s = '%s'", key, value);
```

Comment:
This cannot handle the case of string literal with `'` inside. You should be able to directly construct a `Predicate` using the key and value, then wrap it as a `FilterContext` then pass it into the `JsonIndexReader`

Reply:
Please add a test to verify it works with a string value with `'` inside